### PR TITLE
Split DB to separate pod

### PR DESCRIPTION
**Pull request by romayalon**
### Explain the changes
1. added statefulset-db, and moved the relevant parts from statefuleset-core to it.
2. added db-pod and service-db to support the changes.

### Issues: Fixed #xxx / Gap #xxx
1.
### Testing Instructions:
1.

### CODE REVIEW

**Comment by mykaul** [comment]
What's the motivation behind this change?

**Comment by romayalon** [comment]
It's the first step of moving to micro-services.

**Comment by nimrod-becker** [comment]
Adding to Romy, also the first step for multi noobaa core, opening the flows of geo cluster, DR etc.

**Comment by mykaul** [comment]
Excellent - would be great to link this PR to an issue describing the motivation and architecture.

File: pkg/system/phase2_creating.go [hunk 1]
```diff
@@ -75,6 +86,21 @@ func (r *Reconciler) SetDesiredServiceAccount() {
 		`{"kind":"OAuthRedirectReference","apiVersion":"v1","reference":{"kind":"Route","name":"` + r.RouteMgmt.Name + `"}}`
 }
 
+// ReconcileNoobaaDB reconciles NoobaaDB statefulset
+func (r *Reconciler) ReconcileNoobaaDB() error {
+	if err := r.ReconcileObject(r.NoobaaDB, r.SetDesiredNoobaaDB); err != nil {
+		return err
+	}
+	// check if mongoDB is ready. if not return error
+	if !util.KubeCheck(r.NoobaaDBPod) {
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]: Why do we need to wait for the pod before we can continue?
There are plenty of things to create in phase2, and better create it all and then wait for all to start serving in phase3.

File: pkg/system/phase2_creating.go [hunk 2]
```diff
@@ -85,54 +111,28 @@ func (r *Reconciler) SetDesiredServiceS3() {
...
 
-// SetDesiredCoreApp updates the CoreApp as desired for reconciling
-func (r *Reconciler) SetDesiredCoreApp() {
-	r.CoreApp.Spec.Template.Labels["noobaa-core"] = r.Request.Name
-	r.CoreApp.Spec.Template.Labels["noobaa-mgmt"] = r.Request.Name
-	r.CoreApp.Spec.Template.Labels["noobaa-s3"] = r.Request.Name
-	r.CoreApp.Spec.Selector.MatchLabels["noobaa-core"] = r.Request.Name
-	r.CoreApp.Spec.ServiceName = r.ServiceMgmt.Name
+// SetDesiredServicedb updates the ServiceS3 as desired for reconciling
+func (r *Reconciler) SetDesiredServicedb() {
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]:
```suggestion
func (r *Reconciler) SetDesiredServiceDB() {
```

File: deploy/internal/statefulset-core.yaml [hunk 3]
```diff
@@ -33,124 +22,95 @@ spec:
...
+            - containerPort: 6443
+            - containerPort: 8080
+            - containerPort: 8443
+            - containerPort: 8444
+            - containerPort: 8445
+            - containerPort: 8446
+            - containerPort: 60100
+          env:
+            - name: MONGODB_URL
+              value: "mongodb://127.0.0.1/nbcore"
```
Comment (guymguym) [inline deploy/internal/statefulset-core.yaml]:
```suggestion
              value: "mongodb://noobaa-db-0.noobaa-db/nbcore"
```

File: pkg/system/phase2_creating.go [hunk 4]
```diff
@@ -210,7 +207,75 @@ func (r *Reconciler) SetDesiredCoreApp() {
...
+			}
+			for j := range c.Env {
+				switch c.Env[j].Name {
+				// case "ENDPOINT_FORKS_NUMBER":
+				// 	c.Env[j].Value = "1" // TODO recalculate
+				case "AGENT_PROFILE":
+					c.Env[j].Value = r.SetDesiredAgentProfile(c.Env[j].Value)
+
+				case "MONGODB_URL":
+					c.Env[j].Value = "mongodb://noobaa-db-0.noobaa-db/nbcore"
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]:
```suggestion
					c.Env[j].Value = "mongodb://" + r.NoobaaDB.Name + "-0." + r.NoobaaDB.Name + "/nbcore"
```

File: pkg/system/phase2_creating.go [hunk 5]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
 }
+
+// fixPvs make changes related to pvs if an upgrade process is needed
+func (r *Reconciler) fixPvs() {
+
+	// declare the old pvc
+	oldPvc := &corev1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{Kind: "PersistentVolumeClaim"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "db-" + options.SystemName + "-core-0",
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]:
```suggestion
			Name:      "db-" + r.CoreApp.Name + "-0",
```

File: pkg/system/phase2_creating.go [hunk 6]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+			Namespace: options.Namespace,
+		},
+	}
+	// declare the new pvc
+	newPvc := &corev1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{Kind: "PersistentVolumeClaim"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "db-" + options.SystemName + "-db-0",
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]:
```suggestion
			Name:      "db-" + r.NoobaaDB.Name + "-0",
```

File: pkg/system/reconciler.go [hunk 7]
```diff
@@ -66,8 +66,11 @@ type Reconciler struct {
 	NooBaa              *nbv1.NooBaa
 	ServiceAccount      *corev1.ServiceAccount
 	CoreApp             *appsv1.StatefulSet
+	NoobaaDB            *appsv1.StatefulSet
```
Comment (guymguym) [inline pkg/system/reconciler.go]: Use capital B for NooBaa -
```suggestion
	NooBaaDB            *appsv1.StatefulSet
```

File: pkg/system/phase2_creating.go [hunk 8]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
 	util.Panic(err)
 	return string(profileBytes)
 }
+
+// fixPvs make changes related to pvs if an upgrade process is needed
+func (r *Reconciler) fixPvs() {
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]: I suggest a more specific name
```suggestion
func (r *Reconciler) UpgradeRenameDBVolume() {
```

File: pkg/system/phase2_creating.go [hunk 9]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+		// delete old noobaa-core statefulset
+		oldSts := &appsv1.StatefulSet{
+			TypeMeta: metav1.TypeMeta{Kind: "StatefulSet"},
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "noobaa-core",
+				Namespace: options.Namespace,
+			},
+		}
+		util.KubeCheck(oldSts)
+		util.KubeDelete(oldSts)
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]: Sounds a bit weird that this is done before setting the Retain reclaim policy on the pv, no?

File: pkg/system/phase2_creating.go [hunk 10]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+
+		newAccessMode := oldPvc.Spec.AccessModes
+		newResources := oldPvc.Spec.Resources
+
+		util.KubeCheck(pv)
+		pv.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimRetain
+		pv.Spec.ClaimRef = nil
+		util.KubeUpdate(pv)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go]: add kubecheck and return err if reclaimPolicy is didn't change

File: pkg/system/phase2_creating.go [hunk 11]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+				Namespace: options.Namespace,
+			},
+		}
+		util.KubeCheck(oldSts)
+		util.KubeDelete(oldSts)
+
+		newAccessMode := oldPvc.Spec.AccessModes
+		newResources := oldPvc.Spec.Resources
+
+		util.KubeCheck(pv)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go]: check if !kubecheck ...

File: pkg/system/phase2_creating.go [hunk 12]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+
+		// declare the Spec of the new pvc
+		newPvc.Spec = corev1.PersistentVolumeClaimSpec{
+			VolumeName:       pvName,
+			StorageClassName: r.NooBaa.Spec.DBStorageClass,
+			AccessModes:      newAccessMode,
+			Resources:        newResources,
+		}
+		// create the new pvc
+		newPvcCreated := util.KubeCreateSkipExisting(newPvc)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go]: if ! KubeCreateSkipExisting return err

File: pkg/system/phase2_creating.go [hunk 13]
```diff
@@ -353,3 +418,88 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+			util.KubeUpdate(oldPvc)
+			util.KubeDelete(oldPvc)
+		}
+	}
+	r.Logger.Infof("Version is updated.")
+	// fix reclaim policy to retain
+	if pvName != "" && util.KubeCheck(pv) {
+		if pv.Spec.PersistentVolumeReclaimPolicy == "Retain" {
+			pv.Spec.PersistentVolumeReclaimPolicy = corev1.PersistentVolumeReclaimDelete
+			util.KubeUpdate(pv)
```
Comment (romayalon) [inline pkg/system/phase2_creating.go]: check !kubeUpdate

File: pkg/system/reconciler.go [hunk 14]
```diff
@@ -66,8 +66,11 @@ type Reconciler struct {
 	NooBaa              *nbv1.NooBaa
 	ServiceAccount      *corev1.ServiceAccount
 	CoreApp             *appsv1.StatefulSet
+	NooBaaDB            *appsv1.StatefulSet
+	NoobaaDBPod         *corev1.Pod
```
Comment (guymguym) [inline pkg/system/reconciler.go]: Remove `NoobaaDBPod`...

File: pkg/system/phase2_creating.go [hunk 15]
```diff
@@ -353,3 +405,105 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+			VolumeName:       pvName,
+			StorageClassName: r.NooBaa.Spec.DBStorageClass,
+			AccessModes:      newAccessMode,
+			Resources:        newResources,
+		}
+		// create the new pvc
+		if !util.KubeCreateSkipExisting(newPvc) {
+			return fmt.Errorf("New PVC could not be created")
+		}
+		if !util.RemoveFinalizer(oldPvc, "kubernetes.io/pvc-protection") {
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]:
```suggestion
// delete old noobaa-core statefulset
oldSts := &appsv1.StatefulSet{
    TypeMeta: metav1.TypeMeta{Kind: "StatefulSet"},
    ObjectMeta: metav1.ObjectMeta{
        Name:      "noobaa-core",
        Namespace: options.Namespace,
    },
}
util.KubeDelete(oldSts)


if util.KubeCheck(oldSts) {
    return fmt.Errorf("OLD PVC is not deleted")
}
if util.KubeCheck(oldPvc) {
    return fmt.Errorf("OLD PVC is not deleted")
}
```

File: pkg/system/phase2_creating.go [hunk 16]
```diff
@@ -210,7 +194,75 @@ func (r *Reconciler) SetDesiredCoreApp() {
...
+			[]corev1.LocalObjectReference{}
+	} else {
+		podSpec.ImagePullSecrets =
+			[]corev1.LocalObjectReference{*r.NooBaa.Spec.ImagePullSecret}
+	}
+	if r.NooBaa.Spec.Tolerations != nil {
+		podSpec.Tolerations = r.NooBaa.Spec.Tolerations
+	}
+
+	r.CoreApp.Spec.VolumeClaimTemplates = nil
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]: BTW - Why is this needed?

File: pkg/system/phase2_creating.go [hunk 17]
```diff
@@ -353,3 +405,109 @@ func (r *Reconciler) SetDesiredAgentProfile(profileString string) string {
...
+	newPvc := &corev1.PersistentVolumeClaim{
+		TypeMeta: metav1.TypeMeta{Kind: "PersistentVolumeClaim"},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "db-" + r.NooBaaDB.Name + "-0",
+			Namespace: options.Namespace,
+		},
+		Spec: oldPvc.Spec,
+	}
+	util.KubeCreateSkipExisting(newPvc)
+	if !util.KubeCheck(newPvc) {
```
Comment (guymguym) [inline pkg/system/phase2_creating.go]: Do you think we should add a short sleep here to avoid a restart of reconcile in most cases?
